perun-core/src/main/java/cz/metacentrum/perun/core/impl/modules/attributes/urn_perun_user_attribute_def_virt_loa.java

```diff
@@ -51,4 +64,51 @@ public AttributeDefinition getAttributeDefinition() {
 		return attr;
 	}
 
+	@Override
+	public List<AuditEvent> resolveVirtualAttributeValueChange(PerunSessionImpl sess, AuditEvent message) throws InternalErrorException, AttributeNotExistsException, WrongAttributeAssignmentException {
+
+		List<AuditEvent> resolvingMessages = new ArrayList<>();
+		if (message == null) return resolvingMessages;
+
+		if (message instanceof UserExtSourceAddedToUser) {
+			resolvingMessages.addAll(resolveEvent(sess, ((UserExtSourceAddedToUser) message).getUser()));
+		} else if (message instanceof UserExtSourceRemovedFromUser) {
+			resolvingMessages.addAll(resolveEvent(sess, ((UserExtSourceRemovedFromUser) message).getUser()));
+		} else if (message instanceof UserExtSourceUpdated) {
+			try {
+				resolvingMessages.addAll(resolveEvent(sess, sess.getPerunBl().getUsersManagerBl().getUserById(
+						sess, ((UserExtSourceUpdated) message).getUserExtSource().getUserId())));
+			} catch (UserNotExistsException e) {
+				throw new ConsistencyErrorException("User associated with updated UserExtSource no longer exists while resolving virtual attribute value change.", e);
+			}
+		}
+		return resolvingMessages;
+
+	}
+
+	/**
+	 * Resolve and create new auditer messages about LOA attribute change based on current attribute value.
+	 *
+	 * @param sess PerunSession
+	 * @param user User to resolve LoA messages
+	 * @return List of new messages or empty list
+	 * @throws InternalErrorException
+	 * @throws AttributeNotExistsException
+	 * @throws WrongAttributeAssignmentException
+	 */
+	private List<AuditEvent> resolveEvent(PerunSessionImpl sess, User user) throws InternalErrorException, AttributeNotExistsException, WrongAttributeAssignmentException {
```

Comment:
Why is output of this method the list of audit events if only one message is always generated?

Reply:
Fixed, I do return single ``AuditEvent`` now.